File: Recursion/python/1780.py
n = int(input())
arr = [list(map(int,input().split())) for _ in range(n)]
result = [0, 0, 0]

def search (x,y,n):
    num = arr[x][y]
    flag = False
    for i in range(n):
        for j in range(n):
            if arr[x+i][y+j] != num:
                flag = True
                break
        if flag:
            break
    if flag:
        next = n//3
        search(x,y,next)
        search(x,y+next,next)
        search(x,y+next*2,next)
        search(x+next,y,next)
        search(x+next,y+next,next)
        search(x+next,y+next*2,next)
        search(x+next*2,y,next)
        search(x+next*2,y+next,next)
        search(x+next*2,y+next*2,next)
    else:
        result[num+1] += 1
search(0,0,n)
print("\n".join(map(str,result)))

File: Recursion/javascript/1780.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");
const n = Number(input[0]);
const arr = [];
input.slice(1, input.length - 1).forEach((a) => {
  arr.push(a.split(" ").map(Number));
});

const result = [0, 0, 0];

const check = (x, y, n) => {
  const NUM = arr[x][y];
  let flag = false;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (arr[x + i][y + j] !== NUM) {
        flag = true;
        break;
      }
    }
    if (flag) {
      break;
    }
  }
  if (flag) {
    const next = n / 3;
    check(x, y, n / 3);
    check(x, y + next, n / 3);
    check(x, y + next * 2, n / 3);
    check(x + next, y, n / 3);
    check(x + next, y + next, n / 3);
    check(x + next, y + next * 2, n / 3);
    check(x + next * 2, y, n / 3);
    check(x + next * 2, y + next, n / 3);
    check(x + next * 2, y + next * 2, n / 3);
  } else {
    result[NUM + 1] += 1;
  }
};
check(0, 0, n);
console.log(result.join("\n"));
